### PR TITLE
feat(components): refactor loading

### PR DESCRIPTION
**Pull request by sxzz**
- supports directive binding props / 支持指令传参数
```vue
<div v-loading="{ body: true }">123</div>
```
- supports reactive options / 支持传递响应式选项
```ts
const text = ref('123')
const background = ref('#123')
const loading = ElLoading.service({ text, background })
```
- add `before-close` prop
- add `closed` prop

---

Depreacte Plan: remove `element-loading-xxx` in the next minor release.

- https://github.com/element-plus/element-plus/pull/3113

### CODE REVIEW

**Comment by sxzz** [comment]
Please reviewer re-run unit test (can't pass tests in the morning）

File: packages/components/loading/src/directive.ts [hunk 1]
```diff
@@ -1,56 +1,99 @@
-import Loading from './index'
+import { isRef, ref } from 'vue'
+import { Loading } from './service'
+import type { LoadingOptions } from './types'
+import type { LoadingInstance } from './loading'
+import type { Directive, DirectiveBinding, UnwrapRef } from 'vue'
```
Comment (JeremyWuuuuu) [inline packages/components/loading/src/directive.ts:5]: **Suggestion:**  Move line#5 to line#3

File: packages/components/loading/src/directive.ts [hunk 2]
```diff
@@ -1,56 +1,99 @@
...
-    target: binding.modifiers.fullscreen ? null : el,
-    body: !!binding.modifiers.body,
-    visible: true,
-    lock: !!binding.modifiers.lock,
-  })
+
+  const getBindingProp = <K extends keyof LoadingOptions>(
+    key: K
+  ): LoadingOptions[K] =>
+    typeof binding.value === 'object' ? binding.value[key] : undefined
```
Comment (JeremyWuuuuu) [inline packages/components/loading/src/directive.ts:26]: **Suggestion:** can we use `isObject` under `@element-plus/utils/util` or `@vue/shared`?

**Comment by sxzz** [comment]
Done.

File: packages/components/loading/src/directive.ts [hunk 3]
```diff
@@ -1,56 +1,100 @@
...
+    key: K
+  ): LoadingOptions[K] =>
+    isObject(binding.value) ? binding.value[key] : undefined
+  const resolveExpression = (key: any) => {
+    const data = (isString(key) && vm?.[key]) || key
+    if (data) return ref(data)
+    else return data
+  }
+
+  const text = getBindingProp('text') || el.getAttribute('element-loading-text')
```
Comment (JeremyWuuuuu) [inline packages/components/loading/src/directive.ts:34]: **Suggestion:** encapsulate these into function:

```ts
const getProp = (name: string) => {
  return getBindingProp(name) || el.getAttribute(`element-loading-${hyphen(name)}`)
}
```

Comment (sxzz) [reply packages/components/loading/src/directive.ts:34]: 1. `svgViewBox` -> `element-loading-svg-view-box` not same.
2. I planed to remove `element-loading-xxx` in the next minor release. So let's keep it as is.

**Review by jw-foss — APPROVED** [review]

**Pull request merged**